sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala

```diff
@@ -30,8 +36,156 @@ case class GpuJsonToStructs(
     timeZoneId: Option[String] = None)
     extends GpuUnaryExpression with TimeZoneAwareExpression with ExpectsInputTypes
         with NullIntolerant {
+  
+  private def cleanAndConcat(input: cudf.ColumnVector): (cudf.ColumnVector, cudf.ColumnVector) ={
+    withResource(cudf.Scalar.fromString("{}")) { emptyRow =>
+      val stripped = withResource(cudf.Scalar.fromString(" ")) { space =>
+        input.strip(space)
+      }
+      withResource(stripped) { stripped =>
+        val isNullOrEmptyInput = withResource(input.isNull) { isNull =>
+          val isEmpty = withResource(stripped.getCharLengths) { lengths =>
+            withResource(cudf.Scalar.fromInt(0)) { zero =>
+              lengths.lessOrEqualTo(zero)
+            }
+          }
+          withResource(isEmpty) { isEmpty =>
+            isNull.binaryOp(cudf.BinaryOp.NULL_LOGICAL_OR, isEmpty, cudf.DType.BOOL8)
+          }
+        }
+        closeOnExcept(isNullOrEmptyInput) { _ =>
+          withResource(isNullOrEmptyInput.ifElse(emptyRow, stripped)) { cleaned =>
+            withResource(cudf.Scalar.fromString("\n")) { lineSep =>
+              withResource(cudf.Scalar.fromString("\r")) { returnSep =>
+                withResource(cleaned.stringContains(lineSep)) { inputHas =>
+                  withResource(inputHas.any()) { anyLineSep =>
+                    if (anyLineSep.isValid && anyLineSep.getBoolean) {
+                      throw new IllegalArgumentException("We cannot currently support parsing " +
+                          "JSON that contains a line separator in it")
+                    }
+                  }
+                }
+                withResource(cleaned.stringContains(returnSep)) { inputHas =>
+                  withResource(inputHas.any()) { anyReturnSep =>
+                    if (anyReturnSep.isValid && anyReturnSep.getBoolean) {
+                      throw new IllegalArgumentException("We cannot currently support parsing " +
+                          "JSON that contains a carriage return in it")
+                    }
+                  }
+                }
+              }
+              (isNullOrEmptyInput, cleaned.joinStrings(lineSep, emptyRow))
+            }
+          }
+        }
+      }
+    }
+  }
+
+  // Process a sequence of field names. If there are duplicated field names, we only keep the field
+  // name with the largest index in the sequence, for others, replace the field names with null.
+  // Example:
+  // Input = [("a", StringType), ("b", StringType), ("a", IntegerType)]
+  // Output = [(null, StringType), ("b", StringType), ("a", IntegerType)]
+  private def processFieldNames(names: Seq[(String, DataType)]): Seq[(String, DataType)] = {
+    val existingNames = Set[String]()
+    names.foldRight(Seq[(String, DataType)]())((elem, acc) => {
+      val (name, dtype) = elem
+      if (existingNames(name)) (null, dtype)+:acc else {existingNames += name; (name, dtype)+:acc}})
```

Comment:
Thanks for explanation! It will be easier to read with more line breaks IMO. 

nit: If we move existingNames into fold we can get away without a mutable Set.
```suggestion
    val zero = (Set.empty[String], Seq.empty[(String, DataType)])
    val (_, res) = names.foldRight(zero) { case ((name, dtype), (existingNames, acc)) =>
      if (existingNames(name)) {
        (existingNames, (null, dtype) +: acc)
      } else {
        (existingNames + name, (name, dtype) +: acc)
      }
    }
    res
```

Reply:
Thank you very much! Changes updated.